src/test/java/uk/nhs/digital/nhsconnect/lab/results/translator/mapper/ObservationMapperTest.java

```diff
@@ -211,21 +259,218 @@ void testMapToTestGroupsAndResultsRange() {
+                .containsExactly("url-organization", "url-practitioner"));
+    }
+
+    @Test
+    void testMapNoPerformers() {
+        final var message = new Message(List.of(
+            "S02+02", // ServiceReportDetails
+            "S06+06", // InvestigationSubject
+            "GIS+N", // LabResult
+            "INV+MQ+c:911::d", // LaboratoryInvestigation
+            "RFF+ASL:1" // Reference
+        ));
+
+        final var observations = mapper.mapToObservations(message, mockPatient, emptyMap(), null, null);
+
+        assertThat(observations).hasSize(1).first()
+            .extracting(Observation::getPerformer)
+            .satisfies(performers -> assertThat(performers).isEmpty());
+    }
+
     @Test
     void testCanMapMultipleTestResults() {
-        final Message message = new Message(List.of(
+        final var message = new Message(List.of(
             "S02+02",
             "S06+06",
             "GIS+N",
             "INV+MQ+c:911::d",
+            "RFF+ASL:1",
+            "GIS+N",
+            "INV+MQ+c:911::d",
+            "RFF+ASL:1",
             "GIS+N",
             "INV+MQ+c:911::d",
+            "RFF+ASL:1"
+        ));
+
+        final var observations = mapper.mapToObservations(message, mockPatient, emptyMap(), mockOrganization,
+            mockPractitioner);
+
+        assertThat(observations).hasSize(3);
+    }
+
+    @Test
+    void testMapSpecimen() {
+        final var message = new Message(List.of(
+            "S02+02",
+            "S06+06",
+            "S16+16",
+            "SEQ++1",
             "GIS+N",
-            "INV+MQ+c:911::d"
+            "INV+MQ+c:911::group",
+            "RFF+ASL:1"
         ));
+        var mockSpecimen = mock(Specimen.class);
+        when(mockSpecimen.getId()).thenReturn("uuid-specimen");
+        when(fullUrlGenerator.generate("uuid-specimen")).thenReturn("url-specimen");
+
+        final var observations = mapper.mapToObservations(message, mockPatient, Map.of(mockSpecimen, "1"),
+            mockOrganization, mockPractitioner);
+
+        assertThat(observations).hasSize(1)
+            .map(Observation::getSpecimen)
+            .map(Reference::getReference)
+            .first()
+            .isEqualTo("url-specimen");
+    }
 
-        final List<Observation> actual = mapper.mapToTestGroupsAndResults(message);
+    @Test
+    void testMapTestGroupsWithBidirectionalReferences() {
+        final var message = new Message(List.of(
+            "S02+02",
+            "S06+06",
+
+            "GIS+N",
+            "INV+MQ+c:911::group",
+            "SEQ++1",
+            "RFF+ASL:1",
+
+            "GIS+N",
+            "INV+MQ+c:911::result",
+            "RFF+ARL:1"
+        ));
+
+        /* TODO: review -> this is messy, but otherwise we have to know in which order the
+            implementation is processing them. Perhaps that's okay, on balance? */
+        final var uuidToUrlMap = Map.of("uuid-1", "url-1", "uuid-2", "url-2");
```

Comment:
I agree with the way you've implemented this. I can't see a better way without knowing the order in which the implementation processes the bidirectional mappings.

Reply:
Then I'll update the comment to briefly explain the strategy. Something like:

```
/* We don't know whether the test group or test result will be processed first,
 * so we don't know which observation is given which UUID. But, we can still
 * check the references by training the fullUrlGenerator with known mappings
 * and then verify the actual ID of one corresponds with the actual reference
 * of the other. */
```

Reply:
Yeah, I like this comment - explains why and what very nicely.

I know we generally prefer self-commenting code, but I think this is a case where we need a comment is warranted.